addons/website_sale/tests/test_website_sale_pricelist.py

```diff
@@ -427,10 +427,8 @@ def setUp(self):
             'company_id': self.company2.id,
             'website_id': False,
         })
-        self.demo_user.partner_id.property_product_pricelist = self.c1_pl
-        # Switch env.user company to create ir.property in company2
-        self.env.user.company_id = self.company2
```

Comment:
@rco-odoo Clearly this changes the `Environment` in-place which makes the `lru_cache` no longer valid, the way I "fixed" it is by using `with_company` instead which uses a brand new `Environment`.

Should we allow in-place modifications of the `Environment`? I.e. should we assume the `Environment` is immutable and any modifications should be done through new envs? (`with_company`, `with_user`, `with_context`, etc?)

Note that `functools.lru_cache` actually provides a `clear_cache` method on the wrapped method so we could invalidate the cache in some cases, the question is whether this should be done manually or through ORM magicks

Reply:
Was there any conclusion regarding this particular issue?

Reply:
@odony We talked about it in person and concluded that the environment shouldn't be modified and should be assumed as immutable (ofc it's not actually immutable unfortunately) since we already have all the approriate `with_*` methods.

Ideally we'd raise an error / log a warning if an Environment is modified in-place but I'm uncertain of its feasibility, maybe we could at least change the documentation to reflect this pitfall?

What's your opinion on this?